README.md
## one-file-pdf - A minimalist PDF generator in &lt;2K lines and 1 file
[![Go Report Card](https://goreportcard.com/badge/github.com/balacode/one-file-pdf)](https://goreportcard.com/report/github.com/balacode/one-file-pdf)
[![Build Status](https://travis-ci.org/balacode/one-file-pdf.svg?branch=master)](https://travis-ci.org/balacode/one-file-pdf)
[![Test Coverage](https://coveralls.io/repos/github/balacode/one-file-pdf/badge.svg?branch=master&service=github)](https://coveralls.io/github/balacode/one-file-pdf?branch=master)
[![Gitter chat](https://badges.gitter.im/balacode/one-file-pdf.png)](https://gitter.im/one-file-pdf/Lobby)
[![godoc](https://godoc.org/github.com/balacode/one-file-pdf?status.svg)](https://godoc.org/github.com/balacode/one-file-pdf)
[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)    

The main idea behind this project was:  
*"How small can I make a PDF generator for it to still be useful for 80% of common PDF generation needs?"*

The result is a single .go file with less than 1999 lines of code, about 400 of which are color and glyph-size constants, and ~350 are comments.

- It's easier to learn about the internals of the PDF format with a small, concise library.
- The current version of the file is indicated in the header (the timestamp).

## Features:  
- The essentials for generating PDF documents, sufficient for common business reports.
- Use all built-in PDF fonts: Courier, Helvetica, Symbol, Times, ZapfDingbats, and their variants
- Specify colo(u)rs by name (144 web colors), HTML codes (#RRGGBB) or RGB value
- Set columns for text (like tab stops on the page)
- Built-in grid option to help measurement and positioning
- Metadata properties: author, creator, keywords, subject and title
- Set the measurement units you want: mm, cm, inches, twips or points
- Draw lines with different thickness
- Filled or outline rectangles, circles and ellipses
- JPEG, GIF and transparent PNG images (filled with specified background color)
- Stream compression can be turned on or off (PDF files normally compress streams to reduce file size, but turning it off helps in debugging or learning about PDF commands)

## Not Yet Supported:  
- Unicode (requires font embedding)
- Font embedding
- PDF encryption
- Paths, curves and complex graphics

## Installation:  

```bash
    go get github.com/balacode/one-file-pdf
```

## Naming Convention:  
All types in are prefixed with PDF for public, and 'pdf' for private types.
The only type you need to use is PDF, while PDFColorNames are left public for reference.

## Hello World:  

```go
package main 

import "fmt"
import "github.com/balacode/one-file-pdf"

func main() {
    fmt.Println(`Generating a "Hello World" PDF...`)

    // create a new PDF using 'A4' page size
    var pdf = pdf.NewPDF("A4")

    // set the measurement units to centimeters
    pdf.SetUnits("cm")

    // draw a grid to help us align stuff (just a guide, not necessary)
    pdf.DrawUnitGrid()

    // draw the word 'HELLO' in orange, using 100pt bold Helvetica font
    // - text is placed on top of, not below the Y-coordinate
    // - you can use method chaining
    pdf.SetFont("Helvetica-Bold", 100).
        SetXY(5, 5).
        SetColor("Orange").
        DrawText("HELLO")

    // draw the word 'WORLD' in blue-violet, using 100pt Helvetica font
    // note that here we use the colo(u)r hex code instead
    // of its name, using the CSS/HTML format: #RRGGBB
    pdf.SetXY(5, 9).
        SetColor("#8A2BE2").
        SetFont("Helvetica", 100).
        DrawText("WORLD!")

    // draw a flower icon using 300pt Zapf-Dingbats font
    pdf.SetX(7).SetY(17).
        SetColorRGB(255, 0, 0).
        SetFont("ZapfDingbats", 300).
        DrawText("a")

    // save the file:
    // if the file exists, it will be overwritten
    // if the file is in use, prints an error message
    pdf.SaveFile("hello.pdf")
} //                                                                        main
```

## Samples:
Click on a sample to see the PDF in more detail.

[!["Hello World!" sample image](demo/samples/hello.png)](demo/samples/hello.pdf)  

[!["Synergy Ipsum" sample image](demo/samples/corporate.png)](demo/samples/corporate.pdf)  

## Changelog:  

These are the most recent changes in the functionality of the package,
not including internal changes which are best seen in the commits history.

**2018-04-14**
- Changed CurrentPage from read-only to read/write property: added SetCurrentPage()
- Created PageCount() read-only property
- Created dingbats() demo to generate `zapf_dingbats_table.pdf`.
  You can use this table to look up the hex code for each icon.
- Changed text encoding from /WinAnsiEncoding to /StandardEncoding

See [changelog.md](./doc/changelog.md) for changes made earlier.

## Roadmap:  

- Achieve 100% test coverage
- Create a unit test for every method
- Unicode support
- Partial font embedding


doc/changelog.md
## Detailed Changelog  

These are the changes in the functionality of the package, 
not including internal changes. Internal changes are are 
best seen in the commits history.  

**2018-MAR-30**  
- **ALTERED API: Removed SetErrorLogger() method**
- **ALTERED API: ToPoints(): added error return value**  
- Initialize PDF automatically, even when NewPDF() wasn't called. The paper size
  is A4, and the units CM by default. To specify a different paper size, use NewPDF().
- No need to add the first page with AddPage(). It is inserted automatically.
- New error handling methods Clean(), Errors(), ErrorInfo() and PullError().
- SetColumnWidths(): can be called without arguments when you need to reset all columns.
- Added various unit tests, for 95% code coverage.
- Fixed text wrapping bug that could cause PDF to freeze.

**2018-MAR-14**
- Added support for color JPEG, GIF and PNG images with transparency blending
- Added all standard A, B, and C paper sizes, and US Tabloid and Ledger
- DrawImage(): added backColor optional parameter (using ...) so you can specify the background color for transparent PNGs
- DrawImage(): changed to draw images down from the Y-coordinate position (below, not above Y)
- Created "PNG Images Demo", which outputs to png_images.pdf
- Created ToColor() function to convert named colors and HTML color codes to RGBA color values
- Created ToUnits() method to convert points to the currently-active units of measurement
- Removed PDFNoPage constant
- Created basic unit tests and test helper functions
- Various internal changes, reducing file length by about 60 lines

**2018-MAR-08**
- New methods DrawCircle(), DrawEllipse(), FillCircle(), FillEllipse()
- New demo demonstrating circles and text wrapping: corporate.pdf ("Synergy Ipsum")
- SetColor(): now allows HTML color values like `"#4C9141 MayGreen"`, ignores the extra chars.
- Log an error when the name of a selected font is unknown.
- Log the specified measurement unit's name when its name is not valid.

**2018-MAR-09**
- Replaced PDFColor type with standard lib's color.RGBA (import "image/color")
- SetColorRGB(): changed parameters from int to uint8
- Changed PDFPageSize and PDFStandardPageSizes to private structures
- SaveFile() now returns `error` instead of `*PDF`, to allow caller to check for IO errors
- SetColumnWidths() is no longer limited to 100 columns
- Font names and color names can be specified with spaces, underscores or '-' delimiting words
- Removed module-global PDFErrorHandler, created SetErrorLogger() to set the handler for each PDF instance


demo/demo.go
// -----------------------------------------------------------------------------
// (c) balarabe@protonmail.com                                      License: MIT
// :v: 2018-04-10 17:21:11 B4F377                                      [demo.go]
// -----------------------------------------------------------------------------

package main

// This demo generates the sample PDF files
// and demonstrates the API of One-File-PDF

import (
	"fmt"
	"strings"

	"github.com/balacode/one-file-pdf"
)

func main() {
	/*1*/ helloWorld()
	/*2*/ corporateIpsum()
	/*3*/ pngImages()
	/*4*/ dingbats()
} //                                                                        main

func helloWorld() {
	fmt.Println(`Generating a "Hello World" PDF...`)
	//
	// create a new PDF using 'A4' page size
	var doc = pdf.NewPDF("A4")
	//
	// you can call AddPage() to add the first page, but it is
	// not required: the first page is added automatically
	// doc.AddPage()
	//
	// set the measurement units to centimeters
	doc.SetUnits("cm")
	//
	// draw a grid to help us align stuff (just a guide, not necessary)
	doc.DrawUnitGrid()
	//
	// draw the word 'HELLO' in orange, using 100pt bold Helvetica font
	// - text is placed on top of, not below the Y-coordinate
	// - you can use method chaining
	doc.SetFont("Helvetica-Bold", 100).
		SetXY(5, 5).
		SetColor("Orange").
		DrawText("HELLO")
	//
	// draw the word 'WORLD' in blue-violet, using 100pt Helvetica font
	// note that here we use the colo(u)r hex code instead
	// of its name, using the CSS/HTML format: #RRGGBB
	doc.SetXY(5, 9).
		SetColor("#8A2BE2").
		SetFont("Helvetica", 100).
		DrawText("WORLD!")
	//
	// draw a flower icon using 300pt Zapf-Dingbats font
	doc.SetXY(7, 17).
		SetColor("Red").
		SetFont("ZapfDingbats", 300).
		DrawText("a")
	//
	// save the file:
	// if the file exists, it will be overwritten
	// if the file is in use, prints an error message
	doc.SaveFile("hello.pdf")
} //                                                                  helloWorld

func corporateIpsum() {
	const FILENAME = "corporate.pdf"
	fmt.Println("Generating sample PDF:", FILENAME, "...")
	var doc = pdf.NewPDF("A4") // create a new PDF using 'A4' page size
	doc.SetUnits("cm")
	//
	// draw the heading
	doc.SetColor("#002FA7 InternationalKleinBlue").
		FillBox(0, 1.5, 21, 1.5).
		SetFont("Helvetica-Bold", 50).
		SetColor("White").
		SetXY(3.5, 2.7).DrawText("Synergy Ipsum")
	//
	// draw the green circle
	doc.SetColor("#74C365 Mantis").FillCircle(21, 21, 10) // x, y, radius
	//
	// draw the left column of text (in a box)
	var col1 = strings.Replace(CorporateFiller1, "\n", " ", -1)
	doc.SetColor("#73C2FB MayaBlue").
		FillBox(0, 4, 10, 15). // xywh
		SetColor("black").
		SetFont("times-roman", 11).
		DrawTextInBox(0.5, 4.5, 9, 15, "LT", col1)
	//
	// draw the right column of text
	var col2 = strings.Replace(CorporateFiller2, "\n", " ", -1)
	doc.SetColor("black").
		SetFont("Times-Italic", 11).
		DrawTextInBox(10.5, 4, 9, 28, "LT", col2)
	//
	// draw the bottom-left box with a checkmark
	doc.SetColor("#EAA221 Marigold").
		FillBox(0, 25, 5, 5). // xywh
		SetFont("zapfdingbats", 50).
		SetColor("white").
		DrawTextInBox(0, 25, 5, 5, "C", string(rune(063)))
	//
	// save the file
	doc.SaveFile(FILENAME)
} //                                                              corporateIpsum

func pngImages() {
	const FILENAME = "png_images.pdf"
	fmt.Println("Generating sample PDF:", FILENAME, "...")
	var doc = pdf.NewPDF("A4")
	doc.SetUnits("cm")
	//
	// draw background pattern
	for x := 0.0; x < doc.PageWidth(); x += 6 {
		for y := 0.0; y < doc.PageHeight(); y += 5 {
			doc.DrawImage(x, y, 5, "../image/gophers.png", "cyan")
		}
	}
	// draw dice
	doc.SetColor("WHITE").FillBox(3.5, 4.5, 14.7, 17).
		//
		DrawImage(4, 5, 5, "../image/dice.png", "WHITE").
		DrawImage(11, 5, 5, "../image/dice.png", "RED").
		//
		DrawImage(4, 10.5, 5, "../image/dice.png", "GREEN").
		DrawImage(11, 10.5, 5, "../image/dice.png", "BLUE").
		//
		DrawImage(4, 16, 5, "../image/dice.png", "BLACK").
		SetFont("Helvetica-Bold", 50).
		SetXY(3, 3).SetColor("#009150").
		DrawText("PNG Image Demo")
	//
	doc.SaveFile(FILENAME)
} //                                                                   pngImages

// dingbats generates a useful table of icon codes for Zapf-Dingbats
func dingbats() {
	const filename = "zapf_dingbats_table.pdf"
	fmt.Println("Generating", filename, "...")
	//
	// create a new PDF using 'A4' page size
	var doc = pdf.NewPDF("A4")
	doc.SetUnits("cm")
	//
	const boxSize = 1.2 // cm
	var x, y = 1.0, 1.0 // cm
	//
	doc.SetFont("Helvetica-Bold", 100)
	doc.SetLineWidth(0.02)
	//
	for row := 0; row < 16; row++ {
		x = 1.0
		for col := 0; col < 16; col++ {
			//
			// draw border around each icon
			doc.SetColor("gray")
			doc.DrawBox(x, y, boxSize, boxSize)
			//
			// draw hex code
			doc.SetColor("dark green")
			doc.SetFont("Helvetica", 7)
			doc.DrawTextInBox(x+0.1, y, boxSize, boxSize, "TL",
				fmt.Sprintf("%02Xh", row*16+col))
			//
			// draw decimmal code
			doc.SetColor("dark violet")
			doc.SetFont("Helvetica", 7)
			doc.DrawTextInBox(x, y, boxSize, boxSize, "TR",
				fmt.Sprintf("%d", row*16+col))
			//
			// this is the right way to use a dingbat code (0-255):
			// (casting int to rune to string won't work expected)
			var code = row*16 + col
			var s = string([]byte{byte(code)})
			//
			// draw the dingbat icon
			doc.SetColor("black")
			doc.SetFont("ZapfDingbats", 20)
			doc.DrawTextInBox(x, y+0.1, boxSize, boxSize, "C", s)
			//
			x += boxSize
		}
		y += boxSize
	}
	doc.SaveFile(filename)
} //                                                                    dingbats

//end
